tests/poll.rs

```diff
@@ -473,6 +473,58 @@ fn poll_ok_after_cancelling_pending_ops() {
     handle.join().expect("unable to join thread");
 }
 
+// This test checks the following reregister constraint:
+// The event source must have previously been registered with this instance
+// of `Poll`, otherwise the behavior is undefined.
+//
+// This test is done on Windows and epoll platforms where reregistering a
+// source without a previous register is defined behavior that fail with an
+// error code.
+//
+// On kqueue platforms reregistering w/o registering works but that's not a
+// test goal, so it is not tested.
+#[test]
+#[cfg(any(target_os = "linux", target_os = "windows"))]
```

Comment:
As `TcpListener` uses `IoSource` this should work on all platforms, see https://github.com/tokio-rs/mio/blob/c5f52893ffd4b641d1a457f0197804e69ad666e4/src/io_source.rs#L246-L266